### DISCUSSION TITLE
Error when adding particles in Fortran

### DISCUSSION THREAD

**Discussion by HoulinLee** [AMReX-Codes/amrex#2751] [Q&A]
Dear all,
I’m a Fortran user of AMReX, and I use the functionalities of Particles in AMReX. In my codes, I consider the ionization collision between particles, which means that new particles will be added during the simulation. I add particles by the following procedure (`PC_ele_particles` is the `ParticleContainer` of electrons and `PC_ion_particles` is the `ParticleContainer` of ions): 
```
type(amrex_particle), pointer :: particles(:)
particles=>PC_ele_particles%get_particles(lev,mfi) 
elec_num=size(particles)
do ll = 1, elec_num
! Determine whether the ionization occurs
if (an ionization collision occurs) then
……
! Add a new electron
call amrex_set_particle_cpu(amrex_get_cpu(),p_new1)
call amrex_set_particle_id(amrex_get_next_particle_id(),p_new1)
call PC_ele_particles%add_particle(lev,mfi,p_new1)
! Add a new ion
call amrex_set_particle_cpu(amrex_get_cpu(),p_new2)
call amrex_set_particle_id(amrex_get_next_particle_id(),p_new2)
call PC_ion_particles%add_particle(lev,mfi,p_new2)
……
enddo
```
However, I found that an error occurs when the frequency of ionization collision is high. There is no specific error message but the codes suddenly stop and produce a file of _Backtrace.0_ to tell me the location where the error occurs. After some tests, I found that any call to the pointer `particles(ll)`, such as `particles(ll)%pos`, will cause a segment fault after the call of `PC_ele_particles%add_particle`. Also, I found the error disappear if I call `PC_ele_particles%add_particle` and `PC_ion_particles%add_particle` outside the cycle of `do ll = 1, elec_num enddo`. So, I wonder if there is a restriction for the use of `add_particle`, such as one could not add particles in the cycle where `particles`, which is a pointer of `amrex_particle`, is used. Thanks.

**Comment by atmyers** [comment]
I think the issue is that `add_particle` invalidates the pointer because it can potentially trigger a resize in the underlying data vector. After calling it, you need to do `particles=>PC_ele_particles%get_particles(lev,mfi)` again before indexing in to `particles` or otherwise using it.

**Reply by HoulinLee** [reply]
Thanks for your reply. I'll try as you suggested.
